File: frontend/src/stores/federation.js
import { defineStore } from 'pinia';
import FederationService from '../services/FederationService';

export const useFederationStore = defineStore('federation', {
  state: () => ({
    peers: [],
    selectedPeer: null,
    federatedEvents: [],
    tradeRequests: [],
    isLoading: false,
    error: null,
  }),
  actions: {
    // --- Peers ---
    async fetchPeers() {
      this.isLoading = true;
      try {
        this.peers = await FederationService.getPeers();
      } catch (error) {
        this.error = error;
        console.error('Failed to fetch peers:', error);
      } finally {
        this.isLoading = false;
      }
    },
    async fetchPeer(peerId) {
      this.isLoading = true;
      try {
        this.selectedPeer = await FederationService.getPeer(peerId);
      } catch (error) {
        this.error = error;
        console.error(`Failed to fetch peer ${peerId}:`, error);
      } finally {
        this.isLoading = false;
      }
    },
    async createPeer(peerData) {
      this.isLoading = true;
      try {
        const newPeer = await FederationService.createPeer(peerData);
        this.peers.push(newPeer);
        return newPeer;
      } catch (error) {
        this.error = error;
        console.error('Failed to create peer:', error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async updatePeer(peerId, peerData) {
      this.isLoading = true;
      try {
        const updatedPeer = await FederationService.updatePeer(peerId, peerData);
        const index = this.peers.findIndex(p => p.id === peerId);
        if (index !== -1) {
          this.peers[index] = updatedPeer;
        }
        if (this.selectedPeer?.id === peerId) {
          this.selectedPeer = updatedPeer;
        }
        return updatedPeer;
      } catch (error) {
        this.error = error;
        console.error(`Failed to update peer ${peerId}:`, error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async deletePeer(peerId) {
      this.isLoading = true;
      try {
        await FederationService.deletePeer(peerId);
        this.peers = this.peers.filter(p => p.id !== peerId);
        if (this.selectedPeer?.id === peerId) {
          this.selectedPeer = null;
        }
      } catch (error) {
        this.error = error;
        console.error(`Failed to delete peer ${peerId}:`, error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },

    // --- Federated Events ---
    async fetchFederatedEvents(skip = 0, limit = 20) {
      this.isLoading = true;
      try {
        this.federatedEvents = await FederationService.getFederatedEvents(skip, limit);
      } catch (error) {
        this.error = error;
        console.error('Failed to fetch federated events:', error);
      } finally {
        this.isLoading = false;
      }
    },
    async fetchFederatedEvent(eventId) {
      this.isLoading = true;
      try {
        const event = await FederationService.getFederatedEvent(eventId);
        return event;
      } catch (error) {
        this.error = error;
        console.error(`Failed to fetch federated event ${eventId}:`, error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async createFederatedEvent(peerId, eventData) {
      this.isLoading = true;
      try {
        const newEvent = await FederationService.createFederatedEvent(peerId, eventData);
        this.federatedEvents.push(newEvent); // Or update relevant list
        return newEvent;
      } catch (error) {
        this.error = error;
        console.error('Failed to create federated event:', error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async updateFederatedEvent(eventId, eventData) {
      this.isLoading = true;
      try {
        const updatedEvent = await FederationService.updateFederatedEvent(eventId, eventData);
        const index = this.federatedEvents.findIndex(e => e.id === eventId);
        if (index !== -1) {
          this.federatedEvents[index] = updatedEvent;
        }
        return updatedEvent;
      } catch (error) {
        this.error = error;
        console.error(`Failed to update federated event ${eventId}:`, error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async deleteFederatedEvent(eventId) {
      this.isLoading = true;
      try {
        await FederationService.deleteFederatedEvent(eventId);
        this.federatedEvents = this.federatedEvents.filter(e => e.id !== eventId);
      } catch (error) {
        this.error = error;
        console.error(`Failed to delete federated event ${eventId}:`, error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },

    // --- Trade Requests ---
    async fetchTradeRequests(skip = 0, limit = 20) {
      this.isLoading = true;
      try {
        this.tradeRequests = await FederationService.getTradeRequests(skip, limit);
      } catch (error) {
        this.error = error;
        console.error('Failed to fetch trade requests:', error);
      } finally {
        this.isLoading = false;
      }
    },
    async fetchTradeRequest(requestId) {
      this.isLoading = true;
      try {
        const request = await FederationService.getTradeRequest(requestId);
        return request;
      } catch (error) {
        this.error = error;
        console.error(`Failed to fetch trade request ${requestId}:`, error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async createTradeRequest(peerId, requestData) {
      this.isLoading = true;
      try {
        const newRequest = await FederationService.createTradeRequest(peerId, requestData);
        this.tradeRequests.push(newRequest); // Or update relevant list
        return newRequest;
      } catch (error) {
        this.error = error;
        console.error('Failed to create trade request:', error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async updateTradeRequest(requestId, requestData) {
      this.isLoading = true;
      try {
        const updatedRequest = await FederationService.updateTradeRequest(requestId, requestData);
        const index = this.tradeRequests.findIndex(r => r.id === requestId);
        if (index !== -1) {
          this.tradeRequests[index] = updatedRequest;
        }
        return updatedRequest;
      } catch (error) {
        this.error = error;
        console.error(`Failed to update trade request ${requestId}:`, error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async deleteTradeRequest(requestId) {
      this.isLoading = true;
      try {
        await FederationService.deleteTradeRequest(requestId);
        this.tradeRequests = this.tradeRequests.filter(r => r.id !== requestId);
      } catch (error) {
        this.error = error;
        console.error(`Failed to delete trade request ${requestId}:`, error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
  },
});

File: backend/app/routers/federation.py
from typing import Annotated, List, Optional
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from app.database import get_db
from app.dependencies import get_current_approved_user, require_permission
from app.models.user import User
from app.schemas.federation import (
    PeeredInstanceCreate, PeeredInstanceUpdate, PeeredInstanceResponse,
    FederatedEventCreate, FederatedEventResponse,
    TradeRequestCreate, TradeRequestResponse
)
from app.services.federation import FederationService

router = APIRouter(prefix="/api/federation", tags=["federation"])

# --- Peers ---
@router.post("/peers", response_model=PeeredInstanceResponse, status_code=status.HTTP_201_CREATED)
async def create_peer(
    data: PeeredInstanceCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    return await service.create_peer(data)

@router.get("/peers", response_model=List[PeeredInstanceResponse])
async def list_peers(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
):
    service = FederationService(db)
    return await service.get_peers(skip, limit)

@router.get("/peers/{peer_id}", response_model=PeeredInstanceResponse)
async def get_peer(
    peer_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    service = FederationService(db)
    peer = await service.get_peer(peer_id)
    if not peer:
        raise HTTPException(status_code=404, detail="Peer not found")
    return peer

@router.patch("/peers/{peer_id}", response_model=PeeredInstanceResponse)
async def update_peer(
    peer_id: int,
    data: PeeredInstanceUpdate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    peer = await service.get_peer(peer_id)
    if not peer:
        raise HTTPException(status_code=404, detail="Peer not found")
    return await service.update_peer(peer, data)

@router.delete("/peers/{peer_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_peer(
    peer_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    peer = await service.get_peer(peer_id)
    if not peer:
        raise HTTPException(status_code=404, detail="Peer not found")
    await service.delete_peer(peer)

# --- Federated Events ---
@router.post("/peers/{peer_id}/events", response_model=FederatedEventResponse, status_code=status.HTTP_201_CREATED)
async def create_federated_event(
    peer_id: int,
    data: FederatedEventCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    peer = await service.get_peer(peer_id)
    if not peer:
        raise HTTPException(status_code=404, detail="Peer not found")
    return await service.create_federated_event(peer_id, data)

@router.get("/events", response_model=List[FederatedEventResponse])
async def list_federated_events(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
):
    service = FederationService(db)
    return await service.get_federated_events(skip, limit)

@router.get("/events/{event_id}", response_model=FederatedEventResponse)
async def get_federated_event(
    event_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    service = FederationService(db)
    event = await service.get_federated_event(event_id)
    if not event:
        raise HTTPException(status_code=404, detail="Federated Event not found")
    return event

@router.patch("/events/{event_id}", response_model=FederatedEventResponse)
async def update_federated_event(
    event_id: int,
    data: FederatedEventCreate, # Using Create schema for simplicity, consider Update schema
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    event = await service.get_federated_event(event_id)
    if not event:
        raise HTTPException(status_code=404, detail="Federated Event not found")
    return await service.update_federated_event(event, data)

@router.delete("/events/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_federated_event(
    event_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    event = await service.get_federated_event(event_id)
    if not event:
        raise HTTPException(status_code=404, detail="Federated Event not found")
    await service.delete_federated_event(event)


# --- Trade Requests ---
@router.post("/peers/{peer_id}/trade-requests", response_model=TradeRequestResponse, status_code=status.HTTP_201_CREATED)
async def create_trade_request(
    peer_id: int,
    data: TradeRequestCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    peer = await service.get_peer(peer_id)
    if not peer:
        raise HTTPException(status_code=404, detail="Peer not found")
    return await service.create_trade_request(peer_id, data)

@router.get("/trade-requests", response_model=List[TradeRequestResponse])
async def list_trade_requests(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
):
    service = FederationService(db)
    return await service.get_trade_requests(skip, limit)

@router.get("/trade-requests/{request_id}", response_model=TradeRequestResponse)
async def get_trade_request(
    request_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    service = FederationService(db)
    request = await service.get_trade_request(request_id)
    if not request:
        raise HTTPException(status_code=404, detail="Trade Request not found")
    return request

@router.patch("/trade-requests/{request_id}", response_model=TradeRequestResponse)
async def update_trade_request(
    request_id: int,
    data: TradeRequestCreate, # Using Create schema for simplicity, consider Update schema
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    request = await service.get_trade_request(request_id)
    if not request:
        raise HTTPException(status_code=404, detail="Trade Request not found")
    return await service.update_trade_request(request, data)

@router.delete("/trade-requests/{request_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_trade_request(
    request_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permission("admin.manage_settings"))],
):
    service = FederationService(db)
    request = await service.get_trade_request(request_id)
    if not request:
        raise HTTPException(status_code=404, detail="Trade Request not found")
    await service.delete_trade_request(request)